lib/utility/src/Logger.cpp

```diff
@@ -125,19 +132,81 @@ Utility::Logger& Utility::Logger::operator<<(char c) {
   return *this;
 }
 
-Utility::Logger& Utility::Logger::operator<<(const char* format) {
-#if defined(UART_LOGGING)
-  std::printf(format);
-#endif
+Logger& Logger::operator<<(const char* format) {
+  Logger::printf(format);
+  return *this;
+}
 
 #if defined(SWO_LOGGING)
-  std::sprintf(buffer, format);
+int swo_printf(const char* format, ...) {
+  va_list arg;
+
+  va_start(arg, format);
+  int num_chars = std::vsprintf(buffer, format, arg);
+  va_end(arg);
+
   char* c = buffer;
 
   while (*c) {
     ITM_SendChar(*c++);
+    num_chars++;
   }
-#endif
 
-  return *this;
+  return num_chars;
+}
+
+// Application callback hook for mbed error events, only needed for SWO since mbed already logs through UART to stderr
+// Partially copied from print_error_report() from mbed_error.c
+void mbed_error_hook(const mbed_error_ctx* error_context) {
```

Comment:
I think this is the wrong place to hook into. what we actually want is for the existing print_error_report to work without needing this extra hook. from a quick look, i think in mbed_error_puts, it calls write(STDERR_FILENO...) so we just need to make sure STDERR_FILENO is SWO interface. more about it here: https://os.mbed.com/docs/mbed-os/v6.7/apis/filehandle.html

Reply:
hmm ok, I was looking into it earlier and I couldn't figure out how to redirect STDERR to SWO. I'll add this to my todo

Reply:
the link shows exactly how to do the redirect!